### ISSUE TITLE
Computing epsilon for Federated Learning with differential Privacy

### ISSUE THREAD

**Issue by codaibk** [google-research/federated#25]
Hi developers,
- I read a few suggestions about computing epsilon for federated learning in differential Privacy but still get some confusions. Therefore I hope you can help me make clear some points.
As implemented in this link (https://github.com/tensorflow/privacy/blob/master/tutorials/mnist_dpsgd_tutorial_keras.py) for centralized model. I also use functions `compute_rdp` and `get_privacy_spent` to compute the epsilon shown in below code:
![image](https://user-images.githubusercontent.com/28720676/110103718-34b01280-7de1-11eb-9339-0a8f8d95a17b.png)

![image](https://user-images.githubusercontent.com/28720676/110103566-092d2800-7de1-11eb-864d-8eac1e2fc2ca.png)
My datasets examples = 3012.
Following your suggestion in this [issue](https://github.com/google-research/federated/issues/18). I calculate the sampling ratio = (clients per round / total clients). My wondering is T "steps"  (number of rounds from federated model training) parameter in `compute_epsilon` function should be = `(FLAGS.total_rounds * 3012) // FLAGS.client_batch_size` or we need to add the ratio also because federated learning step is depend on number of client and client_per_round. If yes, it will be like this:
`eps = compute_epsilon((FLAGS.total_rounds * 3012 *(FLAGS.clients_per_round / number_clients)) // FLAGS.client_batch_size)`
- The other question is that my epsilon value is very high (mean not good secure) even I played with some parameters: number of client, batch size, multiplier noise, different kind of T steps like mentioned above but it doesn't improve much.... So I am doubt about my algorithm for computing the epsilon.
Do you have suggestion? or could you guys please update your code with calculating the compute epsilon in your [code](https://github.com/google-research/federated/blob/master/differential_privacy/emnist/run_federated.py). So everyone will be clear about it.
Thanks.

**Comment by kairouzp** [comment]
With respect to your first question, you should set T = `FLAGS.total_rounds`. T is the number of federated rounds, which is captured by `FLAGS.total_rounds`. 

With respect to your second question, there are a few things you need to be aware of. (1) The sampling probability needs to be small enough for you to get a meaningful epsilon. How many clients do you have per round? If it's around 50, then you should get a reasonable sampling probability. (2) The noise multiplier has to be around 1 for you to get reasonable epsilon. What noise multiplier are you suing?

**Comment by codaibk** [comment]
> r you to get a meaningful epsilon. How many clients do you have per round?

Thanks for your quick reply, I think my sampling probability is small enough (2/100= 0.02). I use 2 clients/100 clients per round. The noise multiplier  I play with : 0.1,0.2,0.3, 0.5 (for 0.5 my accuracy result is already bad, so I don't play it more). Could you update your [code](https://github.com/google-research/federated/blob/master/differential_privacy/emnist/run_federated.py) with calculating epsilon? I think it will be quick? But it will helpful for everyone wants to follow.

**Comment by kairouzp** [comment]
It seems that you may be calculating the sampling probability wrong. The sampling probability is equal to the number of clients per round divided by the total number of clients in the dataset. If you are choosing 2 clients per round and your dataset has around 3,000 clients, then the sampling probability should be 2/3000. Having said that, applying user-level differential privacy with only 2 clients per round is extremely challenging. This explains why you can't go beyond a noise multiplier of 0.5. Did you try to increase the number of clients per round? Can you try using 50? That should allow you to scale up the noise multiplier to something larger than 0.5 (hopefully closer to 1). 

With respect to adding how we compute epsilon, this is a good suggestion. I will discuss it internally.  For now though, you can use the following.

``` 
from tensorflow_privacy.privacy.analysis.rdp_accountant import compute_rdp  

def get_eps_gaussian(q, noise_multiplier, steps, target_delta, orders):
  rdp = compute_rdp(q=q, noise_multiplier=noise_multiplier, steps=steps, orders=orders)
  eps, delta, _ = get_privacy_spent(orders, rdp, target_delta=target_delta)
  return eps

orders = [2, 3, 4, 5, 6, 7, 8, 10, 11, 12, 14, 16, 20, 24, 28, 32, 64, 256]
number_clients = 3012 # Set this to the number of clients in the dataset you are using
q= FLAGS.clients_per_round / number_clients 
noise_multiplier = FLAGS.noise_multiplier 
steps = FLAGS.total_rounds
target_delta = 1e-5

eps = get_eps_gaussian(, noise_multiplier, steps, target_delta, orders)

```

**Comment by codaibk** [comment]
> It seems that you may be calculating the sampling probability wrong. The sampling probability is equal to the number of clients per round divided by the total number of clients in the dataset. If you are choosing 2 clients per round and your dataset has around 3,000 clients, then the sampling probability should be 2/3000. Having said that, applying user-level differential privacy with only 2 clients per round is extremely challenging. This explains why you can't go beyond a noise multiplier of 0.5. Did you try to increase the number of clients per round? Can you try using 50? That should allow you to scale up the noise multiplier to something larger than 0.5 (hopefully closer to 1).
> 
> With respect to adding how we compute epsilon, this is a good suggestion. I will discuss it internally. For now though, you can use the following.
> 
> ```
> from tensorflow_privacy.privacy.analysis.rdp_accountant import compute_rdp  
> from tensorflow_privacy.privacy.analysis.rdp_accountant import get_privacy_spent
> 
> def get_eps(q, noise_multiplier, steps, target_delta, orders):
>   rdp = compute_rdp(q=q, noise_multiplier=noise_multiplier, steps=steps, orders=orders)
>   eps, delta, _ = get_privacy_spent(orders, rdp, target_delta=target_delta)
>   return eps
z
> 
> orders = [2, 3, 4, 5, 6, 7, 8, 10, 11, 12, 14, 16, 20, 24, 28, 32, 64, 256]
> number_clients = 3012 # Set this to the number of clients in the dataset you are using
> q= FLAGS.clients_per_round / number_clients 
> noise_multiplier = FLAGS.noise_multiplier 
> steps = FLAGS.total_rounds
> target_delta = 1e-5
> 
> eps = get_eps(q, noise_multiplier, steps, target_delta, orders)
> ```

Thanks for your reply. My dataset includes 3012 images (not 3012 clients) which separated into 100 clients. Therefore my sampling probability is correct with 2/100 (2 clients per round). Anyway, I will check with larger number of clients per round as you mentioned. Thanks for your suggestion.
